src/client.cpp

```diff
@@ -280,6 +290,36 @@ void CClient::OnNewConnection()
     //### TODO: END ###//
 }
 
+void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
+{
+    // Upon receiving a new client list, we have to reset oldGain and newGain
+    // entries for unused channels. This ensures that a disconnected channel
+    // does not leave behind wrong cached gain values which would leak into
+    // any new channel which reused this channel id.
+    int   iNumConnectedClients = vecChanInfo.Size();
+    bool  bChanIdInUse;
+    for ( int iId = 0; iId < MAX_NUM_CHANNELS; iId++ )
+    {
+        bChanIdInUse = false;
+        for ( int j = 0; j < iNumConnectedClients; j++ )
+        {
+            if ( vecChanInfo[j].iChanID == iId )
```

Comment:
I don't quite like the performance characteristics here. This is `O(MAX_NUM_CHANNELS²)` as worst-case (and runs on every channel connect/disconnect). I can't think of a more performant way though and the GUI code uses a similar construct, so I guess it's fine...?

https://github.com/jamulussoftware/jamulus/blob/fa204ad4f3025b7b2a1a8c1130ad998784faeb4a/src/audiomixerboard.cpp#L1213-L1217

Reply:
Basically for each channel you're searching where in vecChanInfo it is and if it's in use (=found) you do not reset it?
Maybe we could do something with sorting, but haven't thought it through

Reply:
Note: I do not have a full understanding of the code, but would something like that work (probably we'd need to take care of the array sizes)

boolean[] inUse = new boolean[MAX_NUM_CHANNELS];
// get all channels which are in use from vacChanInfo[i]
for (int i = 0;i<inUse.length;i++) {
  inUse[vecChanInfo[i].iChanID] = true;
}

// set all channels which are not in use to gain 1
for (int i = 0;i<OldGain.length;i++) {
  if (!inUse[i]) {
    oldGain[i] = newGain[i] = 1;
  }
}

Reply:
That sounds good, thanks! We could probably even skip the initialization loop. That'd make it `O(2n)=O(n)` again, I guess. Will update the PR later.

Reply:
Yeah. Not sure what the default values of a boolean array in C++ are...

Reply:
I'd still like to do a performance test.

Reply:
Ok. The O(n) approach is around 100 (ms?), the O(n²) is fluctuating around 300+ (ms) by a quick test with < 40 clients (but the running time should stay roughly constant, at least for my approach), it remains to test that it works as expected ;-) - and might also be implemented at the other place since I think the O(n) code is cleaner anyway.

Reply:
Actually, the code you linked looks overly complicated. Probably worth refactoring...

Reply:
> Yeah. Not sure what the default values of a boolean array in C++ are...

Um, I was wrong -- seems like one must initialize explicitly: https://stackoverflow.com/questions/1920430/c-array-initialization/1920481#1920481